nemo/collections/llm/inference/base.py

```diff
@@ -0,0 +1,102 @@
+from megatron.core.inference.common_inference_params import CommonInferenceParams
+from megatron.core.inference.engines.mcore_engine import MCoreEngine
+from megatron.core.inference.model_inference_wrappers.gpt.gpt_inference_wrapper import GPTInferenceWrapper
+from megatron.core.inference.model_inference_wrappers.inference_wrapper_config import InferenceWrapperConfig
+from megatron.core.inference.text_generation_controllers.simple_text_generation_controller import (
+    SimpleTextGenerationController,
+)
+from megatron.core.models.gpt.gpt_model import GPTModel as MCoreGPTModel
+from pytorch_lightning.trainer.states import TrainerFn
+
+import nemo.lightning as nl
+from nemo.lightning import io
+from nemo.lightning.pytorch.strategies.megatron_strategy import MegatronStrategy
+from nemo.lightning.pytorch.strategies.utils import RestoreConfig
+
+
+# We need this wrapper since mcore generate uses tokenizer.detokenize, tokenizer.tokenize to encode and decode prompts
+class MCoreTokenizerWrappper:
+    def __init__(self, tokenizer):
+        self.tokenizer = tokenizer
+        self.eod = tokenizer.eod
+        self.vocab_size = tokenizer.vocab_size
+
+    def detokenize(self, tokens):
+        return self.tokenizer.ids_to_text(tokens)
+
+    def tokenize(self, prompt):
+        return self.tokenizer.text_to_ids(prompt)
+
+
+# TODO: Move to lightning Fabric API.
+def _setup_trainer_and_restore_model(path: Path, trainer: nl.Trainer, model: pl.LightningModule):
+    assert isinstance(trainer.strategy, MegatronStrategy), "Only MegatronStrategy is supported for trainer.strategy."
+    restore_config = RestoreConfig(
+        path=path,
+        load_model_state=True,
+        load_optim_state=False,
+    )
+    trainer.strategy.restore_config = restore_config
+    trainer.ckpt_path = None
+    trainer.strategy.connect(model)
+    if trainer.strategy.launcher is not None:
+        trainer.strategy.launcher.launch(lambda: None, trainer=trainer)
+    trainer.strategy.setup_environment()
+
+    if not model.state_dict():
+        model.configure_model()
+
+    trainer.state.fn = TrainerFn.TESTING
+    trainer.strategy.setup_megatron_parallel(trainer=trainer)
+    trainer.strategy.trainer = trainer
+    trainer.strategy.selective_restore()
+
+
+def setup_model_and_tokenizer(
+    path: Path, trainer: Optional[nl.Trainer] = None, params_dtype: torch.dtype = torch.bfloat16
+) -> tuple[MCoreGPTModel, MCoreTokenizerWrappper]:
+    model: io.TrainerContext = io.load_context(path=path, subpath="model")
+    trainer = trainer or io.load_context(path=path, subpath="trainer")
+    _setup_trainer_and_restore_model(path=path, trainer=trainer, model=model)
+
+    mcore_model = model.module.module.module
+    inference_wrapped_model = GPTInferenceWrapper(
+        mcore_model,
+        InferenceWrapperConfig(
+            hidden_size=mcore_model.config.hidden_size,
+            params_dtype=params_dtype,
+            inference_batch_times_seqlen_threshold=1000,
```

Comment:
Do you just want to make it a static value or make this a user given parameter ? Its actually sometimes really much faster to set this to the highest value possible depending on your config.

Reply:
Yeah I can make it user defined.

Reply:
Addressed in https://github.com/NVIDIA/NeMo/pull/10471/commits/6cd5e50fd210d0195244b0d5ecdf71ac2635c40c